pandas/core/base.py

```diff
@@ -93,7 +93,7 @@ def _ensure_type(self: T, obj) -> T:
 
         Used by type checkers.
         """
-        assert isinstance(obj, type(self)), type(obj)
+        assert issubclass(type(obj), type(self)), type(obj)
```

Comment:
The purpose of ensure_type is to ensure that we have a type and not None for methods with an inplace parameter where the return type is Optional. does this work?
```suggestion
        assert obj is not None
```

Reply:
will probably also need to change
`def _ensure_type(self: T, obj) -> T:` to `def _ensure_type(self, obj: Optional[T]) -> T:`

Reply:
If @simonjayhawkins 's suggestion works, I think that would be the mest fool-proof, though I'm not sure I understand the issue.

Another thing we probably should consider is never return multiple types, so e.g. no ``Optional[FrameOrSeries]`` until we get to python 3.8 as the lowest python version. If we did that ``_ensure_type`` could be pulled from the code base.